plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerContentProvider.java

```diff
@@ -44,197 +41,157 @@
-	implements PropertyChangeListener, IProjectCache.IProjectCacheListener {
-	
+public class OpenShiftExplorerContentProvider implements ITreeContentProvider {
 	private static final List<String> TERMINATED_STATUS = Arrays.asList("Complete", "Failed", "Error", "Cancelled");
-	private static final IProjectCache cache = new OpenShiftProjectCache();
-	
+
+	private OpenshiftUIModel model;
+	private IElementListener listener;
+	private StructuredViewer viewer;
+	private Map<Object, BaseExplorerContentProvider.LoadingStub> stubs= new HashMap<Object, BaseExplorerContentProvider.LoadingStub>();
+
 	public OpenShiftExplorerContentProvider() {
-		super();
-		cache.addListener(this);
+		this(new OpenshiftUIModel());
 	}
-	
-	@Override
-	protected void handleConnectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
-		if (!(connection instanceof Connection)) {
-			return;
+
+	public OpenShiftExplorerContentProvider(OpenshiftUIModel model) {
+		this.model = model;
+		listener = new IElementListener() {
+
+			@Override
+			public void elementChanged(IOpenshiftUIElement<?> element) {
+				if (element instanceof OpenshiftUIModel) {
+					refreshViewer(ConnectionsRegistrySingleton.getInstance());
+				} else {
+					refreshViewer(element);
+				}
+			}
+		};
+		model.addListener(listener);
+	}
+
+	protected void refreshViewer(Object element) {
+		if (viewer != null) {
+			viewer.refresh(element);
 		}
-		if (ConnectionProperties.PROPERTY_REFRESH.equals(property)){
-			if(newValue instanceof Connection) {
-				cache.flushFor((Connection) newValue);
+	}
+	
+	protected void asyncExec(Runnable r) {
+		if (viewer != null) {
+			Control control = viewer.getControl();
+			if (control != null && !control.isDisposed()) {
+				control.getDisplay().asyncExec(r);
 			}
-			refreshViewer(newValue);
 		}
 	}
 
+	@Override
+	public void dispose() {
+		model.removeListener(listener);
+		model.dispose();
+	}
+
 	/**
-	 * Called to obtain the root elements of the tree viewer,
-	 * which should be Connections
+	 * Called to obtain the root elements of the tree viewer, which should be
+	 * Connections
 	 */
 	@Override
-	public Object[] getExplorerElements(final Object parentElement) {
-		if(parentElement instanceof ConnectionsRegistry){
-			ConnectionsRegistry registry = (ConnectionsRegistry) parentElement;
-			return registry.getAll(Connection.class).toArray();
+	public Object[] getElements(Object inputElement) {
+		if (inputElement instanceof ConnectionsRegistry) {
+			return model.getConnections().toArray();
 		} else {
 			return new Object[0];
 		}
 	}
 	
+	private void handleLoadingException(Object parentElement, Throwable e) {
+		LoadingStub stub = makeStub(parentElement);
+		stub.add(e);
+		asyncExec(()-> refreshViewer(stub));
+	}
+
+	private LoadingStub makeStub(Object parentElement) {
+		LoadingStub stub = stubs.get(parentElement);
+		if (stub == null) {
+			stub= new LoadingStub();
+			stubs.put(parentElement, stub);
```

Comment:
I think you are correct. handleException() will be called from a background thread, so this is not clean. Bugger! I don't think it will have negative effects, though.